### DISCUSSION TITLE
Should vcpkg's Windows binaries statically link to the CRT?

### DISCUSSION THREAD

**Discussion by BillyONeal** [microsoft/vcpkg#19083] [Q&A]
Currently, vcpkg's Windows bootstrap downloads a `vcpkg.exe` that requires that the VC Redist be installed first. The rationale for this is that users of vcpkg must have installed Visual Studio first to do anything useful, and Visual Studio's installer will install the VC Redist.

We ran into an internal-to-Microsoft situation where a customer is doing something strange where they don't have the Redist installed but would still like to use vcpkg. We are considering changing the binary to use static linking in support of those customers, but don't want to take the ~30% size increase to vcpkg.exe without a nonzero number of normal vcpkg users who would benefit.

Is this something that would help folks? (For example it seems MinGW folks may like such a change?)

Tagging some folks who look like they made MinGW edits recently:

@dg0yt
@fairlyblank
@autoantwort
@longnguyen2004
@d12fk

**Comment by autoantwort** [comment]
I personally don't need it because I already have VS installed. Otherwise I really don't care about the 30% increased file size as long as the executable is not larger as idk 20 MB.

**Reply by BillyONeal** [reply]
We're talking like 3.4MB -> 4.6MB or something like that, so you're good there.

**Reply by Neumann-A** [reply]
hmm I get 3.4 -> 3.7MB with clang-cl and LTO enabled.  (Im still wondering why vcpkglib is like ~160MB)

**Reply by ras0219-msft** [reply]
Does `clang-cl` + LTO result in a binary that doesn't show `vcruntime140` and `msvcp140` in `dumpbin /dependents`? I just wonder because a 300kb increase does sound an awful lot like normal -O3 + LTO optimizations.

**Reply by Neumann-A** [reply]
```
E:\src\AllProjects\vcpkg\buildtrees\vcpkg\x64-windows-llvm-static-rel>dumpbin vcpkg.exe /dependents
Microsoft (R) COFF/PE Dumper Version 14.29.30038.1
Copyright (C) Microsoft Corporation.  All rights reserved.


Dump of file vcpkg.exe

File Type: EXECUTABLE IMAGE

  Image has the following dependencies:

    KERNEL32.dll
    ADVAPI32.dll
    SHELL32.dll
    VERSION.dll
    WINHTTP.dll
    bcrypt.dll

  Summary

        1000 .00cfg
        B000 .data
        1000 .gehcont
       18000 .pdata
       92000 .rdata
        2000 .reloc
        1000 .rsrc
      2E9000 .text
        1000 .tls
        1000 .voltbl
        1000 _RDATA
```
and yes i used vcpkg to build vcpkg but was a bit disappointed since vcpkg is missing a install target and i had to manually run ninja within vcpkg env. triplet/toolchain at https://github.com/Neumann-A/my-vcpkg-triplets

**Comment by dg0yt** [comment]
I don't have VS installed, but I guess the Redist came on to the PC via other packages. I wouldn't mind if I would need to install the Redist separately.
(Components that don't come with the OS are an issue for GPL apps, but that's not the case for vcpkg itself.)

**Comment by longnguyen2004** [comment]
Personally, I wouldn't mind the 1.2MB (~30%) file size increase, since it's pretty much negligible compared to current Internet speed (10Mbps minimum for most places). However this can be useful for MinGW toolchain distributors who want to bundle vcpkg with their toolchain, since everything will "just work".

**Reply by BillyONeal** [reply]
Can you point at an example of someone doing that?

**Reply by dg0yt** [reply]
If I were a mingw toolchain distributor, I would probably build vcpkg from source.

**Reply by ras0219-msft** [reply]
The advantage of using our "blessed" binary is that it's signed and so will generally interact much more favorably with AV software, which is something that's difficult to replicate as an OSS mingw toolchain packager.

**Reply by longnguyen2004** [reply]
> If I were a mingw toolchain distributor, I would probably build vcpkg from source.

Updating using `bootstrap-vcpkg.bat` still uses Microsoft's build

**Reply by dg0yt** [reply]
I didn't want to start another discussion. I'm fine with the blessed binary, and I know how to build from source. I just want to add an opinion on relevance from an open source toolchain perspective.

BillyONeal's question was more like:
Are there examples of (MinGW) toolchain distributors who want to bundle vcpkg with their toolchain?

**Reply by BillyONeal** [reply]
Right. I just want to find one user (just one) who isn't the internal windows build lab to justify taking the size hit to static link.
